### PR TITLE
check that a separator is found for psbt inputs, outputs, and global map

### PR DESCRIPTION
**Pull request by achow101**
Currently it doesn't make sure that a separator was found so PSBTs missing a trailing separator would still pass. This fixes that and adds a test case for it.

It really only makes sense to check for the separator for the output maps as if an input or global map was missing a separator, the fields following it would be interpreted as belonging to the previous input or global map. However I have added the check for those two anyways to be consistent.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Is this for 0.17.1 backport?

File: src/script/sign.h [hunk 1]
```diff
@@ -300,14 +300,18 @@ struct PSBTInput
...
             // Read
             std::vector<unsigned char> key;
             s >> key;
 
             // the key is empty if that was actually a separator byte
             // This is a special case for key lengths 0 as those are not allowed (except for separator)
-            if (key.empty()) return;
+            if (key.empty()) {
+                found_sep = true;
+                break;
```
Comment (jimmysong) [inline src/script/sign.h]: What's the rationale for breaking here instead of throwing the error right away?

File: src/script/sign.h [hunk 2]
```diff
@@ -300,14 +300,18 @@ struct PSBTInput
...
     inline void Unserialize(Stream& s) {
         // Read loop
+        bool found_sep = false;
         while(!s.empty()) {
             // Read
             std::vector<unsigned char> key;
             s >> key;
 
             // the key is empty if that was actually a separator byte
             // This is a special case for key lengths 0 as those are not allowed (except for separator)
```
Comment (jimmysong) [inline src/script/sign.h]: Comment may need updating?

File: src/script/sign.h [hunk 3]
```diff
@@ -475,14 +483,18 @@ struct PSBTOutput
...
-            if (key.empty()) return;
```
Comment (jimmysong) [inline src/script/sign.h]: same as above, comment may need updating?

File: src/script/sign.h [hunk 4]
```diff
@@ -602,14 +618,18 @@ struct PartiallySignedTransaction
...
         // Read global data
+        bool found_sep = false;
         while(!s.empty()) {
             // Read
             std::vector<unsigned char> key;
             s >> key;
 
             // the key is empty if that was actually a separator byte
             // This is a special case for key lengths 0 as those are not allowed (except for separator)
-            if (key.empty()) break;
```
Comment (jimmysong) [inline src/script/sign.h]: comment may need updating?

File: src/script/sign.h (hunk 1 above)
Comment (achow101) [reply src/script/sign.h]: We want to throw the error if we *don't* get here.

File: src/script/sign.h (hunk 2 above)
Comment (achow101) [reply src/script/sign.h]: No

**Comment by sipa** [comment]
utACK 4fb3388db95f408566e43ebb9736842cfbff0a7d. I think it should go into 0.17.1.

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14377/commits/4fb3388db95f408566e43ebb9736842cfbff0a7d

**Pull request merged**